File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/7-Minimum Cut/minimum_cut.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Given a graph, compute a cut with the fewest number of crossing edges.

Format of graph file:
[number of edges]   (The vertex ID starts from 1.)
[end1_id, end2_id]
[end1_id, end2_id]
"""

__author__ = 'Ziang Lu'

from undirected_graph import AdjacencyList


def main():
    with open('graph_info.txt', 'rt') as f:
        # Construct the graph
        graph = AdjacencyList()

        # Add the vertices
        n_vtx = int(f.readline())
        for vtx_id in range(1, n_vtx + 1):
            graph.add_vtx(new_vtx_id=vtx_id)
        # Add the edges
        for line in f.readlines():
            ends = line.split(' ')
            graph.add_edge(end1_id=int(ends[0]), end2_id=int(ends[1]))

        # Compute a minimum cut
        print('Minimum cut: %d' % graph.compute_minimum_cut())


if __name__ == '__main__':
    main()

File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/7-Minimum Cut/MinimumCut.java
/**
 * Given a graph, compute a cut with the fewest number of crossing edges.
 *
 * Format of the graph file:
 * [number of vertices]   (The vertex ID starts from 1.)
 * [end1ID end2ID]
 * [end1ID end2ID]
 * ...
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import undirected_graph.AdjacencyList;

public class MinimumCut {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        Scanner scanner = null;
        try {
            scanner = new Scanner(new File("graph_info.txt"), "latin1");

            // Construct the graph
            AdjacencyList graph = new AdjacencyList();
            // Add the vertices
            int nVtx = Integer.parseInt(scanner.nextLine());
            for (int vtxID = 1; vtxID <= nVtx; ++vtxID) {
                graph.addVtx(vtxID);
            }
            // Add the edges
            while (scanner.hasNextLine()) {
                String[] ends = scanner.nextLine().split(" ");
                graph.addEdge(Integer.parseInt(ends[0]), Integer.parseInt(ends[1]));
            }

            // Compute a minimum cut
            System.out.println("Minimum cut: " + graph.computeMinimumCut());
        } catch (FileNotFoundException ex) {
            System.out.println("Cannot find the file.");
        } finally {
            if (scanner != null) {
                scanner.close();
            }
        }
    }

}
